Add support for OpenIDConnect token refresh

pshchelo opened the pull request:
Creatively adapted from kubernetes-client/python-base.

Adds new setup extra 'oidc' to install required OAuth-related libraries.

Tested with local Keycloak installation.

pshchelo commented:
should I add proper tests, with mocked out oauth dependencies?

pshchelo commented:
rebased, black-ified, poetry-fied

hjacobs commented:
@pshchelo sorry for the late response. In the meantime there was some refactoring to make the HTTP adapter more pluggable (#56). Can you adapt your PR accordingly?

Thanks!

pshchelo commented:
Yes, sure, will do.
We'd really like to see this in pykube, as we have openid in our k8s, and usual python-kubernetes client is not for humans :-) and has its own bug with keycloak (PR is on review there but is stalled)

hjacobs commented:
@pshchelo thanks!

pshchelo commented:
rebased, conflicts resolved fixed. again, tested against our local keycloak, token is refreshed and the config file is updated.

@hjacobs plz review
Somehow I'm not sure how to properly update poetry.lock in such scenario. Only leave the libs I've added in the patch?

hjacobs commented:
Sorry for the delayed response. I will have a look tomorrow.

hjacobs commented on pykube/config.py:
```diff
@@ -173,6 +173,12 @@ def users(self):
                     BytesOrFile.maybe_set(u, "client-key", self.kubeconfig_file)
+                    if "auth-provider" in u:
+                        BytesOrFile.maybe_set(
+                            u["auth-provider"]["config"],
```
the config key might not be present (would raise KeyError)

hjacobs commented:
@pshchelo can you provide me some instructions on how to test this locally with a test setup (e.g. Keycloak + kind or Minikube)?

hjacobs commented on pykube/http.py:
```diff
@@ -71,17 +80,96 @@ def _auth_gcp(self, request, token, expiry, config):
+        else:
+            verify = None
+        # TODO add timeouts
+        response = requests.get(
```
please add a timeout like your TODO already suggests :smirk:

hjacobs commented on pykube/http.py:
```diff
@@ -71,17 +80,96 @@ def _auth_gcp(self, request, token, expiry, config):
+            verify=verify,
+        )
+
+        if response.status_code != HTTPStatus.OK:
```
if you want to catch all errors (e.g. connection problems), you would need to wrap it in `try`/`except` --- I assume you want to ignore failures and continue with the old token

pshchelo commented:
> @pshchelo can you provide me some instructions on how to test this locally with a test setup (e.g. Keycloak + kind or Minikube)?

I'll try to come up with something, most probably with OpenID connect thru free tier of Google or Okta.
At my place we deploy k8s clusters pointing to a central internal keycloak instance, so I never deploy it myself :-/

pshchelo commented:
@hjacobs added timeouts (using already configured default http one in the `http.py` module) and global ignore of any possible refresh errors, in which case I rely on retries and just re-use already present `id-token`.

your question on how to test this stuff eventually led to me writing a blog post on that

https://pshchelo.github.io/pykube-oidc-refresh.html

using dev Okta account for OpenID provider, microk8s snap on Ubuntu for simple Kubernetes cluster, and self-written script to get a refresh token (see the post).

If you have any questions, I'll be happy to try and answer :-)

pshchelo commented:
@hjacobs ping ^
